src/PersistentOrderedMap.mo

```diff
@@ -0,0 +1,892 @@
+/// Stable key-value map implemented as a red-black tree with nodes storing key-value pairs.
+///
+/// A red-black tree is a balanced binary search tree ordered by the keys.
+///
+/// The tree data structure internally colors each of its nodes either red or black,
+/// and uses this information to balance the tree during the modifying operations.
+///
+/// Performance:
+/// * Runtime: `O(log(n))` worst case cost per insertion, removal, and retrieval operation.
+/// * Space: `O(n)` for storing the entire tree.
+/// `n` denotes the number of key-value entries (i.e. nodes) stored in the tree.
+///
+/// Note:
+/// * Map operations, such as retrieval, insertion, and removal create `O(log(n))` temporary objects that become garbage.
+///
+/// Credits:
+///
+/// The core of this implementation is derived from:
+///
+/// * Ken Friis Larsen's [RedBlackMap.sml](https://github.com/kfl/mosml/blob/master/src/mosmllib/Redblackmap.sml), which itself is based on:
+/// * Stefan Kahrs, "Red-black trees with types", Journal of Functional Programming, 11(4): 425-432 (2001), [version 1 in web appendix](http://www.cs.ukc.ac.uk/people/staff/smk/redblack/rb.html).
+
+
+import Debug "Debug";
+import I "Iter";
+import List "List";
+import Nat "Nat";
+import O "Order";
+
+// TODO: a faster, more compact and less indirect representation would be:
+// type Map<K, V> = {
+//  #red : (Map<K, V>, K, V, Map<K, V>);
+//  #black : (Map<K, V>, K, V, Map<K, V>);
+//  #leaf
+//};
+// (this inlines the colors into the variant, flattens a tuple, and removes a (now) redundant optin, for considerable heap savings.)
+// It would also make sense to maintain the size in a separate root for 0(1) access.
+
+module {
+
+  /// Node color: Either red (`#R`) or black (`#B`).
+  public type Color = { #R; #B };
+
+  /// Red-black tree of nodes with key-value entries, ordered by the keys.
+  /// The keys have the generic type `K` and the values the generic type `V`.
+  /// Leaves are considered implicitly black.
+  public type Map<K, V> = {
+    #node : (Color, Map<K, V>, (K, V), Map<K, V>);
+    #leaf
+  };
+
+  /// Opertaions on `Map`, that require a comparator.
+  ///
+  /// The object should be created once, then used for all the operations
+  /// with `Map` to maintain invariant that comparator did not changed.
+  ///
+  /// `MapOps` contains methods that require `compare` internally:
+  /// operations that may reshape a `Map` or should find something.
```

Comment:
I think `traverse` doesn't require `comp` (since we keep tree structure) so it's outside of the `MapOps`.
So I think it should be
 ```suggestion
   /// operations that update or search a map.
```

Reply:
I guess that's a UX question - should we put all operations in MapOps, even if they don't require the comparison, just for uniformity? I guess we could also do that later easily enough. Something to think about.

Reply:
>  should we put all operations in MapOps

I also thought about this. We were going to measure its overhead on a benchmark and then decide. If there is no difference I would also put it there.

However for now some modules like `RBTree` or `TrieMap` provide some functionality as a module-function, not as a member-function e.g. `map` or `iter`. So we thought that the current solution is not very inconsistent.

Reply:
We did performance experiments:
1. Comparing inlined internal functions and the current version (all calls through `MapOps`):  https://github.com/serokell/motoko-base/pull/21
2. Comparing calling methods through MapOps and through the `Internal` module passing `comp` manually: https://github.com/serokell/motoko-base/pull/20

The 1st experiment shows that the inlined version has more or less the same performance (<1% difference).
The 2nd one shows that a static function call is faster than a similar one from `MapOps` on 1-4% percent. 

Thus we can say that static calls are less convenient but faster, so probably it makes sense to keep the option ?
We can add the rest of the functions into `MapOps` and make the `Internal` module public.
OR keep the current version as compromised

WDYT?

Reply:
So I guess we decide to put all operations in MapOps and leave Internal private for now at a 1-4% perf hit.
Does the perf hit disappear entirely if we inline the Internal functions in MapOps, or am I misunderstanding?